x-pack/plugins/infra/server/plugin.ts

```diff
@@ -4,30 +4,149 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { Plugin, PluginInitializerContext } from 'src/core/server';
+import { CoreSetup, PluginInitializerContext } from 'src/core/server';
+import { Server } from 'hapi';
+import { schema, TypeOf } from '@kbn/config-schema';
+import { i18n } from '@kbn/i18n';
+import { initInfraServer } from './infra_server';
+import { InfraBackendLibs, InfraDomainLibs } from './lib/infra_types';
+import { FrameworkFieldsAdapter } from './lib/adapters/fields/framework_fields_adapter';
+import { KibanaFramework } from './lib/adapters/framework/kibana_framework_adapter';
+import { InfraKibanaLogEntriesAdapter } from './lib/adapters/log_entries/kibana_log_entries_adapter';
+import { KibanaMetricsAdapter } from './lib/adapters/metrics/kibana_metrics_adapter';
+import { InfraElasticsearchSourceStatusAdapter } from './lib/adapters/source_status';
+import { InfraFieldsDomain } from './lib/domains/fields_domain';
+import { InfraLogEntriesDomain } from './lib/domains/log_entries_domain';
+import { InfraMetricsDomain } from './lib/domains/metrics_domain';
+import { LogEntryCategoriesAnalysis, LogEntryRateAnalysis } from './lib/log_analysis';
+import { InfraSnapshot } from './lib/snapshot';
+import { InfraSourceStatus } from './lib/source_status';
+import { InfraSources } from './lib/sources';
+import { InfraServerPluginDeps } from './lib/adapters/framework';
+import { METRICS_FEATURE, LOGS_FEATURE } from './features';
+import { UsageCollector } from './usage/usage_collector';
+import { InfraStaticSourceConfiguration } from './lib/sources/types';
 
-export class InfraPlugin implements Plugin<InfraSetup> {
-  private readonly initContext: PluginInitializerContext;
+export const config = {
+  schema: schema.object({
+    enabled: schema.maybe(schema.boolean()),
+    query: schema.object({
+      partitionSize: schema.maybe(schema.number()),
+      partitionFactor: schema.maybe(schema.number()),
+    }),
+  }),
+};
 
-  constructor(initContext: PluginInitializerContext) {
-    this.initContext = initContext;
+export type InfraConfig = TypeOf<typeof config.schema>;
+
+export interface KbnServer extends Server {
+  usage: any;
+}
+
+const logsSampleDataLinkLabel = i18n.translate('xpack.infra.sampleDataLinkLabel', {
+  defaultMessage: 'Logs',
+});
+
+export interface InfraPluginSetup {
+  defineInternalSourceConfiguration: (
+    sourceId: string,
+    sourceProperties: InfraStaticSourceConfiguration
+  ) => void;
+}
+
+const DEFAULT_CONFIG: InfraConfig = {
+  enabled: true,
+  query: {
+    partitionSize: 75,
+    partitionFactor: 1.2,
+  },
+};
+
+export class InfraServerPlugin {
+  public config: InfraConfig = DEFAULT_CONFIG;
+  public libs: InfraBackendLibs | undefined;
+
+  constructor(context: PluginInitializerContext) {
+    const config$ = context.config.create<InfraConfig>();
+    config$.subscribe(configValue => {
+      this.config = {
```

Comment:
Do we need to make sure that this will propagate to the libs that consume the config or are plugins allowed not to support reloading of the configuration?

Reply:
Same as above.

Reply:
I'm not sure if I understand what you're asking @weltenwort -- but I don't think NP plugins can read from other plugins' config. So I think the values we're adding here are just for ourselves.

Reply:
I'm referring to the fact that `config$` is an observable, so it can emit different values over time. That could happen, for example if Kibana supported reloading the config file without restart.

Reply:
Right, but I think we're okay because we control our own config values, and Kibana shouldn't be trying to emit different values for our own infra config values. So as long as we don't care about emitting different values for those keys, I _think_ we're fine here.